### ISSUE TITLE
Optimizing multiple accesses to same key in map

### ISSUE THREAD

**Issue by rasky** [golang/go#17133]
I was reviewing some code for performance issues, and the profile pointed me to a function that was accessing the same map element multiple times like this:

```go
data[key].Foo = 1
data[key].Bar.Baz += 4
[...]
```

The profile showed many calls to `runtime.mapaccess1_fast32`. I was wondering if it would be legal for the compiler to optimize this straight code to do only one map access, and then of course if it's worth doing it or not.

**Comment by randall77** [comment]
Yes, this would be awesome.  I've been pondering it.  I think we just want to recognize redundant calls and use the same return value from mapaccess* (the pointer to the value) again.

We probably need a mapaccess variant, "lookup key, plus allocate key->0 if not found" method.

Also good for `m[x] += 5`.

See #5147.

**Comment by dgryski** [comment]
It would be nice if this could optimize the one place I've been able to beat the built-in map with a custom implementation: lookup-and-insert-if-not-present. With strings especially as this can require hashing twice.

**Comment by bradfitz** [comment]
Damian, I think I have an existing bug open for that.

**Comment by cespare** [comment]
It's #5147 that @randall77 linked above.

**Comment by dgryski** [comment]
Oops, thanks.  Sorry for the noise.  Subscribed the the original.

**Comment by rsc** [comment]
@randall77, is the second half of this going to land in Go 1.8, or should we move this issue to Go 1.9?

**Comment by randall77** [comment]
Not sure yet.  Magic 8 ball says "looking iffy".  Leave it here for now.

**Comment by griesemer** [comment]
Too late for 1.10.

**Comment by mdempsky** [comment]
> We probably need a mapaccess variant, "lookup key, plus allocate key->0 if not found" method.

This is what mapassign already does. I believe there's nothing blocking us from compiling `m[k] += 5` into `*mapassign(m, k) += 5` rather than `*mapassign(m, k) = *mapaccess1(m, k) + 5`.

As for recognizing repeated map lookups, could that be handled by CSE if we mapped OMAPINDEX to a high-level SSA Op and later lowered it into a runtime call?

**Comment by randall77** [comment]
Note: this has been implemented for the `m[k] op= v` form.
Punting for the generic form.

**Comment by mvdan** [comment]
This would help the flag package, since `flag.FlagSet.Var` needs to first check its map to reject a duplicate flag declaration, and then insert into the map. The double map access currently costs about 10%; see https://go-review.googlesource.com/c/go/+/435259.

**Comment by prattmic** [comment]
> As for recognizing repeated map lookups, could that be handled by CSE if we mapped OMAPINDEX to a high-level SSA Op and later lowered it into a runtime call?

This idea reminds me of ["Representing Data Collections in an SSA Form"](https://users.cs.northwestern.edu/~simonec/files/Research/papers/MEMORY_CGO_2024.pdf), which explores (to great effect) representing container operations in a compiler SSA backend, which enables all sorts of things like detecting that two map lookups for the same key has the same result, even if there is another map insert of a different key in between.
